lib/msal-common/src/cache/CacheManager.ts

```diff
@@ -280,52 +278,127 @@ export abstract class CacheManager implements ICacheManager {
+     * and builds the account info objects from the matching ID token's claims
+     * @param cachedAccounts
+     * @param accountFilter
+     * @returns Array of AccountInfo objects that match account and tenant profile filters
+     */
+    private buildTenantProfiles(
+        cachedAccounts: AccountEntity[],
+        accountFilter?: AccountFilter
+    ): AccountInfo[] {
+        const validAccounts: AccountInfo[] = [];
+        cachedAccounts.forEach((accountEntity) => {
+            const tenantProfile: AccountInfo | null =
+                this.getAccountInfoForTenantProfile(
+                    accountEntity,
+                    accountFilter
+                );
+            if (tenantProfile) {
+                validAccounts.push(tenantProfile);
+            }
+        });
+        return validAccounts;
+    }
+
+    private getAccountInfoForTenantProfile(
         accountEntity: AccountEntity,
         accountFilter?: AccountFilter
+    ): AccountInfo | null {
+        const tenantProfileFilter: TenantProfileFilter = {
+            localAccountId: accountFilter?.localAccountId,
+            loginHint: accountFilter?.loginHint,
+            name: accountFilter?.name,
+            sid: accountFilter?.sid,
+        };
+        return this.getAccountInfoFromEntity(
+            accountEntity,
+            accountFilter?.tenantId,
+            tenantProfileFilter
+        );
+    }
+
+    private getAccountInfoFromEntity(
+        accountEntity: AccountEntity,
+        targetTenantId?: string,
+        tenantProfileFilter?: TenantProfileFilter
     ): AccountInfo | null {
         const accountInfo = accountEntity.getAccountInfo();
-        const idToken = this.getIdToken(accountInfo);
+
+        const idToken = this.getIdToken(accountInfo, undefined, targetTenantId);
+
         if (idToken) {
             const idTokenClaims = extractTokenClaims(
                 idToken.secret,
                 this.cryptoImpl.base64Decode
             );
 
             if (
-                this.idTokenClaimsMatchAccountFilter(
+                this.idTokenClaimsMatchTenantProfileFilter(
                     idTokenClaims,
-                    accountFilter
+                    tenantProfileFilter
                 )
             ) {
-                accountInfo.idToken = idToken.secret;
-                accountInfo.idTokenClaims = idTokenClaims;
-                return accountInfo;
+                return {
+                    ...accountInfo,
+                    idToken: idToken.secret,
+                    idTokenClaims: idTokenClaims,
+                    localAccountId:
+                        idTokenClaims.oid || accountInfo.localAccountId,
+                    name: idTokenClaims.name || accountInfo.name,
+                    tenantId: idTokenClaims.tid || accountInfo.tenantId,
+                };
             }
         }
         return null;
```

Comment:
I'm a little surprised we return null here even if accountInfo is truthy. Do you happen to know why?

Reply:
The accountInfo object, at the point it's checked for truthiness, only contains the base (cached) account entity data. If there is an ID token that matches the filter, then it gets updated with the ID token data and becomes an actual accoutn info object that represents the account in either the tenant from the filter or the home tenant.

If no ID tokens match the filter, then the tenant profile doesn't actually exist, so we return null. Would you like me to change the method name to make that more obvious?

Reply:
Got it. At least a comment to that effect would be helpful

Reply:
Added some comments